### PR TITLE
fix backward index for gamma beta

### PR DESCRIPTION
**Pull request by MARD1NO**
#### 问题描述：
**Eager下**  layernorm前向结果一致，后向不一致，而且weight,bias的grad和torch的是相反过来

#### 初步解决：
猜测是functional赋梯度赋反了，也确实是weight, bias的梯度赋错了，已经修复

- [x] 遗留问题：输入的梯度依旧不对
- [x] 遗留问题：Autotest

### 关于layernorm和batchnorm的联系
layernorm可以看作是batchnorm的一种变种

以输入(b, c, t)，其layernorm(c, t) 等价于，将输入reshape成 (1, b, c*t) 做batchnorm(num_features=3)

```python
import torch
import torch.nn as nn
import numpy as np

x = np.random.randn(3, 2, 2)

torch_x_tensor1 = torch.Tensor(x)
torch_x_tensor2 = torch.Tensor(x)
torch_x_tensor2 = torch_x_tensor2.reshape((1, 3, -1)) # 1, 3, 4

layernorm = nn.LayerNorm((2, 2))
layernorm_out = layernorm(torch_x_tensor1)
print(layernorm.weight.shape) # its weight should be (2, 2)

print("Layer norm out is: ", layernorm_out)

bn = nn.BatchNorm1d(num_features=3) # its weight should be (3, ) equals to num of channels. 
print(bn.weight.shape)
bn_out = bn(torch_x_tensor2)
bn_out = torch.reshape(bn_out, shape=(3, 2, 2))
print("Bn out is: ", bn_out)
```

但我认为这两者**并不严格等价**，layernorm的weight和bias的形状应该是(2, 2)，而bn的weight是(3, )

### 阅读资料
[cudnn文档](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnBatchNormalizationBackward)

[batchnorm反向传播推导](https://medium.com/@drsealks/batch-normalisation-formulas-derivation-253df5b75220)

#### 目前进展：
1. 和喧哥确认，之前lazy实现GPT，和tensorflow，pytorch是能够对齐的
2. 根据chengpeng，liaoxingyu的反馈，在graph跑layernorm也是没问题的

所以还是能够认为以前的这套实现没有问题，可能还是eager，functional这块儿有些问题

### CODE REVIEW

**Comment by yuanms2** [comment]
这个发现很有意思，有没有其他同事能一起确认一下

**Comment by yuanms2** [comment]
查看了一下git blame， 是蔡晟航从基于别的同事开发的版本迁移过来的，当时也发现了计算结果和tf 有一定的差别：

“目前测试发现 epsilon 较大的时候，和 tf 的误差也会较大，甚至达到 1e-1，旧版本也是如此”

https://github.com/Oneflow-Inc/oneflow/pull/2781

**Comment by MARD1NO** [comment]
> 这个发现很有意思，有没有其他同事能一起确认一下

更新了下目前能够确认的内容，目前有guoran，yinggang，depeng在帮忙一起看

**Comment by MARD1NO** [comment]
无意间在python端，改变了gamma 和 beta的传入方式

虽然输入不对齐，但是梯度更新正确了

后续定位到在传入layernorm_param_grad的时候，把gamma传成了beta，修复后能与torch对齐

**Comment by wyg1997** [comment]
还有个问题先记录一下：LayerNorm module 的 forward 中只是用到了 self.normalized_shape 的长度来计算出 norm_axis，并没有真正去 check 成员的 shape 是否对得上。在 TensorDescInfer 时才 CHECK 出错，报错信息也不够清晰，bug 修复后需要注意一下。

File: oneflow/core/autograd/gradient_funcs/layer_norm.cpp [hunk 1]
```diff
@@ -84,13 +82,16 @@ Maybe<void> LayerNorm::Capture(LayerNormCaptureState* ctx, const TensorTuple& in
...
+
+  ctx->has_gamma_diff = ctx->scale && inputs.at(1)->requires_grad();
+  const int beta_index = ctx->scale + 1;
+  ctx->has_beta_diff = ctx->center && inputs.at(beta_index)->requires_grad();
+
   ctx->has_normalized_diff = ctx->scale && inputs.at(0)->requires_grad();
   if (ctx->has_gamma_diff || ctx->has_normalized_diff) {
-    ctx->gamma_index = ctx->SaveTensorForBackward(inputs.at(gamma_index));
+    ctx->gamma_index = ctx->SaveTensorForBackward(inputs.at(1));  // save gamma.
   }
```
Comment (yuanms2) [inline oneflow/core/autograd/gradient_funcs/layer_norm.cpp:93]: 这里 inputs.at(1） 算不算硬编码

File: oneflow/core/autograd/gradient_funcs/layer_norm.cpp [hunk 2]
```diff
@@ -115,28 +116,35 @@ Maybe<void> LayerNorm::Apply(const LayerNormCaptureState* ctx, const TensorTuple
...
     if (ctx->has_gamma_diff) {
       inputs.push_back(saved_tensors.at(ctx->normalized_index));  // normalized
     }
+
     const auto& results = JUST(OpInterpUtil::Dispatch<TensorTuple>(*param_grad_op, inputs));
-    if (ctx->has_beta_diff) { in_grads->at(1) = results->at(0); }
+
     if (ctx->has_gamma_diff) {
-      in_grads->at(ctx->has_beta_diff + 1) = results->at(ctx->has_beta_diff);
+      in_grads->at(1) =
```
Comment (yuanms2) [inline oneflow/core/autograd/gradient_funcs/layer_norm.cpp:130]: at(1) 算不算硬编码

File: oneflow/core/autograd/gradient_funcs/layer_norm.cpp [hunk 3]
```diff
@@ -115,28 +116,35 @@ Maybe<void> LayerNorm::Apply(const LayerNormCaptureState* ctx, const TensorTuple
         begin_params_axis, ctx->has_beta_diff, ctx->has_gamma_diff, ctx->has_normalized_diff,
```
Comment (hjchen2) [inline oneflow/core/autograd/gradient_funcs/layer_norm.cpp:116]: 把LayerNormParamGradOp也顺便迁移到functional吧

Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/layer_norm.cpp:116]: functional LayerNormParamGradOp的输出顺序就和输入保持一致，(dgamma, dbeta)

File: oneflow/core/autograd/gradient_funcs/layer_norm.cpp (hunk 2 above)
Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/layer_norm.cpp:130]: 现在输入和输出都是有序的，这个顺序是在op expr build的时候就确定下来的，这里硬编码没什么问题，效率更高

File: oneflow/core/autograd/gradient_funcs/layer_norm.cpp (hunk 3 above)
Comment (wyg1997) [reply oneflow/core/autograd/gradient_funcs/layer_norm.cpp:116]: 这里每个参数都是有和无两种可能，functor 里就维护每一种可能吗？还有 weight 和 bias 是不是一定是绑定的，affine 为 True 时它们就一定都存在，affine 为 False 就都不存在，这样可以排除一些可能

Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/layer_norm.cpp:116]: pytorch提供了哪些，我们也提供就行了，不用都来一遍

**Comment by MARD1NO** [comment]
#### 记录一个梯度的问题
有些情况下，还是会出现layernorm输出和梯度与torch不对齐的情况（gpu/cpu都会出现）

以一个例子：
```python
import oneflow as flow
import torch
import numpy as np 

x_np = np.array([[[[-1.83965693, -1.82964566]]]])
print(x_np.shape)
affine = False
device = "cuda"

normalized_shape = (1, 1, 2)

of_x_tensor = flow.Tensor(x_np).to(device)
of_x_tensor.requires_grad = True


of_layernorm = flow.nn.LayerNorm(normalized_shape, eps=1e-5, elementwise_affine=affine).to(device)

of_out = of_layernorm(of_x_tensor)
print("Oneflow out is: ", of_out)
of_out = of_out.sum()
of_out.backward()
print("Of x tensor grad is: ", of_x_tensor.grad)


torch_x_tensor = torch.Tensor(x_np).to(device)
torch_x_tensor.requires_grad = True

torch_layernorm = torch.nn.LayerNorm(normalized_shape, eps=1e-5, elementwise_affine=affine).to(device)

torch_out = torch_layernorm(torch_x_tensor)
print("Pytorch out is: ", torch_out)
torch_out = torch_out.sum()
torch_out.backward()
print("Torch x tensor grad is: ", torch_x_tensor.grad)
```
以下是一些实验和猜测：

#### 实验1
使用最naive的方式，用mean和var拼凑：

```python
import oneflow as flow
# import torch as flow 
import numpy as np 


device = "cpu"
eps = 1e-5
x_np = np.array([[[[-1.83965693, -1.82964566]]]])
of_x_tensor = flow.Tensor(x_np).to(device)
of_x_tensor.requires_grad = True

mean = flow.mean(of_x_tensor, dim=(2, 3), keepdim=True)
var = (flow.var(of_x_tensor, dim=(2, 3), keepdim=True, unbiased=False) + eps).rsqrt()
of_out = (of_x_tensor - mean) * var
print("X out is: ", of_out)

of_out = of_out.sum()
of_out.backward()
print("Grad is: ", of_x_tensor.grad)
```
按照拼凑的方式，有一些结论：
1. torch拼凑出来的，和oneflow拼凑出来的，结果，梯度一致
2. 尽管按照原公式拼凑，结果依旧不等价LayerNorm算子（torch.Layernorm输出为[[[[-0.8428,  0.8428]]]]，拼凑实现输出为[[[[-0.8454,  0.8454]]]])

猜测是：
1. 现有的拼凑方式有问题
2. pytorch的LayerNorm有bug

**Comment by simonJJJ** [comment]
This is because of the different algorithms of LayerNorm between PyTorch and oneflow. The Welford algorithm in PyTorch needs higher float precision, after converting torch tensor to float64 could align with oneflow.

**Comment by MARD1NO** [comment]
> This is because of the different algorithms of LayerNorm between PyTorch and oneflow. The Welford algorithm in PyTorch needs higher float precision, after converting torch tensor to float64 could align with oneflow.

感谢世杰帮助

oneflow, lightseq, deepx_core都采用的是比较naive的计算方式，统计得到sum，和sum_square，最后同步，根据公式
```python
D(X) = E(X^2) - E(X)^2 
```
得到方差

而pytorch采取了welford在线算法，在线更新均值和方差，能够节省空间

在dtype为torch.float64下，得到的结果能够对齐

**Review by hjchen2 — APPROVED** [review]

**Review by wyg1997 — COMMENTED** [review]
在 python forward 最上面加一个 check 吧：

**Review by wyg1997 — APPROVED** [review]
